File: cs/UsimmTraceReader.cs
using System;
using System.Globalization;
using System.IO;

namespace Nijikawa
{
    class UsimmTraceReader : ITraceReader
    {
        private StreamReader Reader { get; set; }

        public UsimmTraceReader(string filename)
        {
            Reader = new StreamReader(filename);
        }

        TraceRecord ITraceReader.NextTraceRecord()
        {
            var line = Reader.ReadLine();
            var words = line.Split(' ');
            if (words.Length != 3 && words.Length != 4)
            {
                throw new ApplicationException("Invalid trace file");
            }
            var record = new TraceRecord();
            record.Prec = uint.Parse(words[0]);
            record.Addr = ulong.Parse(words[2].Substring(2),
                    NumberStyles.HexNumber);
            if (words[1] == "R")
            {
                record.IsWrite = false;
            }
            else if (words[1] == "W")
            {
                record.IsWrite = true;
            }
            else
            {
                throw new ApplicationException("Unknown request type " +
                        words[1]);
            }
            return record;
        }
    }
}

File: rust/usimm_trace_reader.rs
use std::from_str::*;
use io = std::io;
use std::path::PosixPath;

use trace::*;

pub struct UsimmTraceReader {
    priv reader: ~io::ReaderUtil,
}

impl UsimmTraceReader {
    pub fn new(filename: &str) -> UsimmTraceReader {
        UsimmTraceReader {
            reader: (~io::file_reader(&PosixPath(filename)).unwrap())
                    as ~ReaderUtil
        }
    }
}

impl TraceReader for UsimmTraceReader {
    fn next_trace_record(&self) -> TraceRecord {
        let line: ~str = self.reader.read_line();
        let words: ~[&str] = line.split_iter(' ').collect();
        assert!(words.len() == 3 || words.len() == 4);
        TraceRecord {
            addr: from_str(words[2]).unwrap(),
            prec: from_str(words[0]).unwrap(),
            is_write: match words[1] {
                "R" => false,
                "W" => true,
                _ => fail2!("unknown request type {:s}", words[1])
            }
        }
    }
}
